bclri instruction support added.

VasiliyMatr opened the pull request:
This instruction returns rs1 with a single bit cleared at the index specified in shamt. The index is read from
the lower log2(XLEN) bits of shamt.

VasiliyMatr committed:
[mipt-v-2021][VasiliyMatr] :
Added myself to members.md.

VasiliyMatr committed:
Merge branch 'main' into main

pavelkryukov committed:
Merge branch 'main' into main

VasiliyMatr committed:
[mipt-v-2021][bclri] : Added bclri instruction support.

pavelkryukov commented:
Please resolve merge conflicts

VasiliyMatr committed:
Merge branch 'main' into main

VasiliyMatr commented:
Resolved merge conflicts.

pavelkryukov commented on simulator/func_sim/alu.h:
```diff
@@ -285,6 +286,14 @@ struct ALU
+    void bclri( Instr* instr )
+    {
+        size_t index = shamt_imm( instr) & (XLEN - 1);
+        size_t mask = ~(static_cast<size_t>( 1) << index);
```
`size_t` won’t cast to `uint128`. How did you compile it?

VasiliyMatr replied:
Compilation successed and all tests were passed. Are there any problems with 128 bit registers?
Specification contains only RV32 and RV64 encodings for this instruction (https://github.com/riscv/riscv-bitmanip/releases/download/1.0.0/bitmanip-1.0.0-38-g865e7a7.pdf). Should I add 128 bit registers support?

pavelkryukov replied:
On CI, compilation failed on Linux. What is your OS, compiler, Boost version?

pavelkryukov replied:
Oh I see. There is a different error, probably a result of gorci leftover

VasiliyMatr replied:
Few problems appeared afted merge: one line were duplicated (I've fixed it in my local repository copy).
But now there is a problem with orc_b opcode. There is no such opcode definition in opcodes-rv32b file in this repo version (but there were no merge conflicts with opcodes-rv32b file somehow). But this definition present in such file if i do git clone of MIPT-ILab/mipt-mips. How could that be?

VasiliyMatr replied:
There were problems with submodules updates.

pavelkryukov commented on simulator/func_sim/alu.h:
```diff
@@ -285,6 +286,14 @@ struct ALU
+    void bclri( Instr* instr )
+    {
+        size_t index = shamt_imm( instr) & (XLEN - 1);
+        size_t mask = ~(static_cast<size_t>( 1) << index);
```
`size_t` won’t cast to `uint128`. How did you compile it?

pavelkryukov commented on simulator/risc_v/riscv_instr.cpp:
```diff
@@ -358,6 +361,8 @@ static const std::vector<RISCVTableEntry<I>> cmd_desc =
     {'B', instr_min,        execute_min<I>,    OUT_ARITHM, ' ', Imm::NO, { Src::RS1, Src::RS2 },  { Dst::RD }, 0, 32 | 64      },
     {'B', instr_minu,       execute_minu<I>,   OUT_ARITHM, ' ', Imm::NO, { Src::RS1, Src::RS2 },  { Dst::RD }, 0, 32 | 64      },
     {'B', instr_gorc,       execute_gorc<I>,   OUT_ARITHM, ' ', Imm::NO, { Src::RS1, Src::RS2 },  { Dst::RD }, 0, 32 | 64      },
+    {'B', instr_gorci,      execute_gorci<I>,  OUT_ARITHM, '7', Imm::ARITH, { Src::RS1, Src::ZERO }, { Dst::RD }, 0, 32 | 64   },
```
We’ve removed `gorci` instruction, please don’t add it back

pavelkryukov commented on simulator/risc_v/t/unit_test.cpp:
```diff
@@ -355,6 +360,13 @@ TEST_RV64_RR_OP( 13, orc_b, 0x000000FFFF000000, 0x0000008007000000, 0)
+TEST_RV32_IMM_OP( 3, bclri, 0xFFFDFFFF, all_ones<uint32>(), 0x11) // 17 bit cleared
+TEST_RV64_IMM_OP( 1, bclri, 0x7FFFFFFFFFFFFFFF, all_ones<uint64>(), 0x3F) // 63 bit cleared (max bit)
+TEST_RV64_IMM_OP( 2, bclri, 0xFFFFFFFFFFFFFFFE, all_ones<uint64>(), 0x00) // 0 bit cleared (min bit)
+TEST_RV64_IMM_OP( 3, bclri, 0xFFFFFFFFFFFDFFFF, all_ones<uint64>(), 0x11) // 17 bit cleared
```
Can we test what happens if bit was 0 already?

VasiliyMatr committed:
[mipt-v-2021][bclri] : Fixed merge errors.

VasiliyMatr committed:
[mipt-v-2021][bclri] : Added few unit tests for bclri instruction.

VasiliyMatr committed:
[mipt-v-2021][bclri] : Fixed bug with size_t. Now using uint64_t.

pavelkryukov commented on simulator/func_sim/alu.h:
```diff
@@ -285,6 +286,14 @@ struct ALU
+    void bclri( Instr* instr )
+    {
+        uint64_t index = shamt_imm( instr) & (XLEN - 1);
+        uint64_t mask = ~(static_cast<uint64_t>( 1) << index);
```
That won't work for RV128, it will cause overflow. Could you please add tests and fix it? I think you need to use `RegisterUInt` here instead of `uint64_t`?

VasiliyMatr replied:
It looks like there is extra bit left for RV128 shamt. I will add RV128 support, but there are only RV32 and RV64 encodings for this instruction, according to specification.

VasiliyMatr replied:
![image](https://user-images.githubusercontent.com/70326825/142249074-ad7e6698-cdf8-4431-b880-e5b27bf13fda.png)

pavelkryukov commented on simulator/func_sim/alu.h:
```diff
@@ -285,6 +286,14 @@ struct ALU
         instr->v_dst[0] = ( (instr->v_src[0] >> pack_width) | (instr->v_src[1] & (bitmask<T>(pack_width) << pack_width)));
     }
 
+    template<typename T> static
```
Template is not used here, can we remove it?

pavelkryukov commented on simulator/risc_v/riscv_instr.cpp:
```diff
@@ -105,6 +105,7 @@ template<typename I> const auto execute_clz = RISCVALU<I>::template clz<typename
 template<typename I> const auto execute_cpop = RISCVALU<I>::template pcnt<typename I::RegisterUInt>;
 template<typename I> const auto execute_ctz = RISCVALU<I>::template ctz<typename I::RegisterUInt>;
 template<typename I> const auto execute_gorc = RISCVALU<I>::template gorc<typename I::RegisterUInt>;
+template<typename I> const auto execute_bclri = RISCVALU<I>::template bclri<typename I::RegisterUInt>;
```
Please sort alphabethically.

pavelkryukov commented on simulator/risc_v/riscv_instr.cpp:
```diff
@@ -359,6 +360,7 @@ static const std::vector<RISCVTableEntry<I>> cmd_desc =
     {'B', instr_minu,       execute_minu<I>,   OUT_ARITHM, ' ', Imm::NO, { Src::RS1, Src::RS2 },  { Dst::RD }, 0, 32 | 64      },
     {'B', instr_gorc,       execute_gorc<I>,   OUT_ARITHM, ' ', Imm::NO, { Src::RS1, Src::RS2 },  { Dst::RD }, 0, 32 | 64      },
     {'B', instr_orc_b,      execute_orc_b<I>,  OUT_ARITHM, ' ', Imm::NO, { Src::RS1, Src::ZERO }, { Dst::RD }, 0, 32 | 64      },
+    {'B', instr_bclri,      execute_bclri<I>,  OUT_ARITHM, '7', Imm::ARITH, { Src::RS1, Src::ZERO }, { Dst::RD }, 0, 32 | 64   },
```
Please move closer to the similar instructions.

VasiliyMatr committed:
[mipt-v-2021][bclri] : Requested changes are made
Sorted code in riscv_instr.cpp;
Removed template from bclri definition;
Using RegisterUInt type in bclri now;
Added tests for RV128 bclri usage.

pavelkryukov commented on simulator/func_sim/alu.h:
```diff
@@ -285,6 +286,13 @@ struct ALU
+    static void bclri( Instr* instr )
+    {
+        size_t index = shamt_imm( instr) & (XLEN - 1);
+        RegisterUInt mask = ~(static_cast<RegisterUInt>( 1) << index);
```
Let's use `auto` for types. Also, use `lsb_set<RegisterUInt>` to get `1` constant of proper size.

pavelkryukov commented on simulator/risc_v/riscv_instr.cpp:
```diff
@@ -93,9 +93,10 @@ template<typename I> const auto execute_divu = RISCVMultALU<I>::template div<typ
 // B
-template<typename I> const auto execute_bseti = RISCVALU<I>::template bseti<typename I::RegisterUInt>;
 template<typename I> const auto execute_add_uw = RISCVALU<I>::template add_uw<typename I::RegisterUInt>;
+template<typename I> const auto execute_bseti = RISCVALU<I>::template bseti<typename I::RegisterUInt>;
```
Please move after `bset`

VasiliyMatr committed:
[mipt-v-2021][bclri] : Minor changes
Sorted riscv_instr.cpp lines;
Using auto for types in bclri function now;
Using lsb_set to get 1 constant of propper type.
